File: ai/tetris/game.py
from typing import *
import numpy as np
import random

from tetris.shape import Shape
from tetris.block import Block, down, left, right, rot_cw, rot_ccw
from tetris.utils import str_bool_ndarray
from tetris.game_type import GameType

free = False
taken = True


class Game:

    #
    # Private methods
    #

    def __init__(self, type: GameType):
        height = type.height
        width = type.width
        shapes = type.shapes
        if height < 1:
            raise ValueError("Height must be at least 1")
        if width < 1:
            raise ValueError("Width must be at least 1")
        if len(shapes) == 0:
            raise ValueError("At least one shape must be provided")

        assert all(map(lambda shape: shape.width < width, shapes)), "No shapes can be wider than board"

        # Extra space needed on top for placing the next falling shape out of view
        top_extra = max(map(lambda shape: shape.height, shapes))
        # Extra space needed on each side so a shape can be up against the wall
        # without it's matrix (and thus, (x, y) location) passing out of bounds.
        side_extra = max(map(lambda shape: shape.width, shapes))
        # 1 added to height for bottom boarder, so no pieces fall through
        bottom_extra = 1
        board = np.full((height + top_extra + bottom_extra, width + side_extra * 2), free, dtype=bool)

        # Add boarders
        board[-1, :] = taken  # Bottom
        board[:, 0:side_extra] = taken  # Left
        board[:, -side_extra:] = taken  # right

        # Fill fields
        self.score: int = 0
        self.board: np.ndarray = board
        self.block: Block = None
        self.shapes: List[Shape] = shapes
        # N, E, S, W
        self.boarder: Tuple[int, int, int, int] = (top_extra, side_extra, bottom_extra, side_extra)
        self.width: int = width
        self.height: int = height
        self.game_over: bool = False

        # Initialize game
        self.gen_block()

    def __str__(self) -> str:
        only_board = self.get_visible()
        if self.block is None:
            string = str_bool_ndarray(only_board)
        else:
            n, e, s, w = self.boarder
            b = self.block
            y = b.y - n
            x = b.x - w
            y1, y2 = y, y + b.height
            x1, x2 = x, x + b.width
            only_block = np.full((self.height, self.width), False, dtype=bool)
            only_block[y1:y2, x1:x2] |= b.matrix()
            string = str_bool_ndarray(only_block | only_board)

        if self.game_over:
            string = "GAME OVER\n" + string + "\nSCORE: %d" % self.score
        return string

    def clear_full_rows(self) -> None:
        """
        Responsible for turning this:

        |      |
        |[][][]| <-
        |  [][]|
        |[][][]| <-
        |  []  |

        into this:

        |      |
        |      | <-
        |  [][]|
        |      | <-
        |  []  |

        and updating scores
        """
        assert not self.game_over
        visible_rows: np.ndarray = self.get_visible()
        streak = 0
        for row in visible_rows:
            if row.all():
                row[:] = free
                streak += 1
            else:
                streak = 0
            if streak:
                # Some random exponential scoring function I made up
                self.score += 10 * (3 ** streak)

    def drop_floating_rows(self) -> None:
        """
        Responsible for turning this:

        |      |
        |      |
        |  [][]|
        |      | <-
        |  []  |

        into this:

        |      |
        |      |
        |      |
        |  [][]|
        |  []  |
        """
        assert not self.game_over
        visible_rows: np.ndarray = self.get_visible()
        # Keep track of this so we know which row to pull from when filling
        # empty row
        dropped = 0
        height = len(visible_rows)
        for i in reversed(range(height)):
            while i - dropped >= 0 and not visible_rows[i - dropped].any():
                dropped += 1
            if i - dropped < 0:
                return
            else:
                # Swap rows (does nothing if dropped is 0)
                visible_rows[[i, i - dropped]] = visible_rows[[i - dropped, i]]

    def gen_block(self):
        assert not self.game_over
        if self.block is not None:
            raise ValueError("Cannot generate two blocks before the last has "
                             "been stamped")
        n, e, s, w = self.boarder
        next_shape: Shape = random.choice(self.shapes)
        init_x: int = e + self.width // 2 - next_shape.width // 2
        init_y: int = n - next_shape.height
        self.block = Block(next_shape, x=init_x, y=init_y)

    def get_visible(self) -> np.ndarray:
        n, e, s, w = self.boarder
        return np.atleast_2d(self.board[n:-s, w:-e])

    def get_staging(self) -> np.ndarray:
        n, e, s, w = self.boarder
        return np.atleast_2d(self.board[0:n, w:-e])

    def attempt_translation(self,
                            translation: Callable[[Block], Block],
                            otherwise: Callable[[], None] = None):
        assert not self.game_over
        new_block = translation(self.block)
        new_y = new_block.y
        new_x = new_block.x
        y1, y2 = new_y, new_y + new_block.height
        x1, x2 = new_x, new_x + new_block.width
        overlapping_new: np.ndarray = self.board[y1:y2, x1:x2]
        assert overlapping_new.shape == new_block.matrix().shape
        if not (overlapping_new & new_block.matrix()).any():
            self.block = new_block
        elif otherwise is not None:
            otherwise()

    def ensure_not_gameover(self):
        if self.game_over:
            raise ValueError("Unsupported when game is over")

    #
    # Public methods
    #

    def down(self) -> None:
        self.ensure_not_gameover()

        def otherwise():
            # Collision, need to stamp current block into board and generate new
            # one, or end game if stamped block enters staging area
            curr_block = self.block
            curr_x = curr_block.x
            curr_y = curr_block.y
            # Stamp block
            y1, y2 = curr_y, curr_y + curr_block.height
            x1, x2 = curr_x, curr_x + curr_block.width
            self.board[y1:y2, x1:x2] |= curr_block.matrix()
            self.block = None
            # Check for end of game
            self.clear_full_rows()
            self.drop_floating_rows()
            if self.get_staging().any():
                self.game_over = True
            else:
                self.gen_block()

        self.attempt_translation(down, otherwise)

    def left(self) -> None:
        self.ensure_not_gameover()
        self.attempt_translation(left)

    def right(self) -> None:
        self.ensure_not_gameover()
        self.attempt_translation(right)

    def rot_cw(self) -> None:
        self.ensure_not_gameover()
        self.attempt_translation(rot_cw)

    def rot_ccw(self) -> None:
        self.ensure_not_gameover()
        self.attempt_translation(rot_ccw)

File: ui/src/game.js
require("utils");
require("matrix");
require("colors");
require("shape");
require("block");

class Game {
	constructor(type) {
		const board_height = type.height;
		const board_width = type.width;
		const shapes = type.shapes;

		assert(board_height > 0, "Board must be at least 1 block high");
		assert(board_width > 0, "Board must be at least 1 block wide");
		assert(shapes.length > 0, "At least one shape is needed to play");
		shapes.forEach((shape, idx) => {
			assert(shape.width < board_width, "No shapes can be wider than the board.", [idx, board_width])
		});

		// Extra space needed on top for placing the next falling shape out of view
		const top_extra = Math.max(shapes.map(shape => shape.height));
		// Extra space needed on each side so a shape can be up against the wall
		// without it's matrix (and thus, (x, y) location) passing out of bounds.
		const side_extra = Math.max(shapes.map(shape => shape.width));
		// 1 added to height for bottom boarder, so no pieces fall through
		// Doesn't need to be more since block matrices can be out of bounds
		// (because x, y start at upper left of block)
		const bottom_extra = 1;
		// Matrix of colors
		const board = new Matrix(
			board_height + top_extra + bottom_extra,
			// times 2 for 1 for each side
			board_width + side_extra * 2,
			(row_idx, col_idx) =>
				col_idx < side_extra                    || // left boarder
				col_idx >= board_width - side_extra     || // right boarder
				row_idx >= board_height - bottom_extra     // bottom boarder
					// Color boarders black, free space clear
					? Colors.BLACK
					: Colors.CLEAR
		);

		// Initialize fields
		this.score = 0;
		this.board = board;
		this.block = null;
		this.shapes = shapes;
		// N, E, S, W
		this.boarder = [top_extra, side_extra, bottom_extra, side_extra];
		this.n_vis_bound = top_extra;
		this.e_vis_bound = board_width - side_extra - 1;
		this.s_vis_bound = board_height - bottom_extra - 1;
		this.w_vis_bound = side_extra;
		this.width = board_width;
		this.height = board_height;
		this.game_over = false;

		this.gen_block();
	}

	assert_game_running() {
		if (this.game_over) {
			throw "Unsupported when game is over";
		}
	}

	get_visible() {
		return  this.board.slice(
			this.n_vis_bound,
			this.w_vis_bound,
			this.s_vis_bound,
			this.e_vis_bound
		)
	}

	/**
	 * Responsible for turning this:
	 *
	 * |      |
	 * |[][][]| <-
	 * |  [][]|
	 * |[][][]| <-
	 * |  []  |
	 *
	 * into this:
	 *
	 * |      |
	 * |      |
	 * |      |
	 * |  [][]|
	 * |  []  |
	 *
	 * and updating scores
	 */
	drop_rows() {
		this.assert_game_running();
		dropped = 0;
		let visible = this.get_visible();
		let rows = visible.copy().rows;
		for (let idx = rows.length - 1; idx >= 0; idx++) {
			let streak = 0;
			while (idx - dropped >= 0 && all(rows[idx - dropped])) {
				dropped += 1;
				streak += 1;
			}
			if (idx - dropped < 0) {
				visible = visible.fill_row(idx, Colors.CLEAR);
			}
			else {
				visible = visible.swap(idx, idx + dropped);
			}
			if (streak) {
				self.score += 10 * Math.pow(3, streak);
			}
		}
		this.board = this.board.replace(this.n_vis_bound, this.w_vis_bound, visible)
	}

	gen_block() {
		this.assert_game_running();
		if (this.block === null) {
			throw "Cannot generate two blocks before the last has been stamped";
		}
		const e = this.boarder[1];
		const next_shape = pick_random(this.shapes);
		const next_color = pick_random(valid_colors);
		const init_x = e + idiv(this.width, 2) - idiv(next_shape.width, 2);
		const init_y = 0;
		this.block = new Block(next_shape, next_color, 0, init_x, init_y);
	}

	get_staging() {
		return this.board.slice(
			0,
			this.w_vis_bound,
			this.n_vis_bound - 1,
			this.e_vis_bound
		)
	}

	translate(new_block) {
		this.assert_game_running();
		const y1 = new_block.y;
		const x1 = new_block.x;
		const y2 = y1 + new_block.height - 1;
		const x2 = x1 + new_block.width - 1;
		const overlapping_new = this.board.slice(y1, x1, y2, x2);
		assert(overlapping_new.cols === new_block.width);
		assert(overlapping_new.rows === new_block.height);
		if (!any(overlapping_new.and(new_block.matrix()))) {
			self.block = new_block;
			return true;
		}
		else {
			return false;
		}
	}

	down() {
		if (!this.translate(this.block.down())) {
			// Collision, need to stamp current block into board and generate new
            // one, or end game if stamped block enters staging area
			this.board = this.board.replace(this.block.y, this.block.x, this.block.matrix())
			this.block = null;
			this.drop_rows();
			if (any(this.get_staging())) {
				this.game_over = true;
			}
			else {
				this.gen_block();
			}
		}
	}

	left() {
		this.translate(this.block.left());
	}

	right() {
		this.translate(this.block.right());
	}

	rot_cw() {
		this.translate(this.block.rot_cw());
	}

	rot_ccw() {
		this.translate(this.block.rot_ccw());
	}
}

file_loaded("game");
